File: web/src/api/endpoints.ts
import { apiGet, apiPost } from "./client";
import type {
  HistoricalSummaryResponse,
  LastEntriesResponse,
  MonthDataResponse,
  MutationResponse,
  UpdateEntryFields,
} from "./types";

export function getMonthData(month?: string | null): Promise<MonthDataResponse> {
  return apiGet<MonthDataResponse>("monthData", { month: month || undefined });
}

export function getHistoricalSummary(): Promise<HistoricalSummaryResponse> {
  return apiGet<HistoricalSummaryResponse>("historicalSummary");
}

export function getLastEntries(n = 10): Promise<LastEntriesResponse> {
  return apiGet<LastEntriesResponse>("lastEntries", { n });
}

export function updateEntry(
  row: number,
  fields: UpdateEntryFields,
): Promise<MutationResponse> {
  return apiPost<MutationResponse>("updateEntry", { row, fields });
}

export function deleteEntry(row: number): Promise<MutationResponse> {
  return apiPost<MutationResponse>("deleteEntry", { row });
}

File: app/lib/api/endpoints.dart
// Spec: docs/specs/api/endpoints.md

import '../core/types.dart';
import 'client.dart';

class ApiEndpoints {
  final ApiClient _client;

  ApiEndpoints(this._client);

  Future<MonthDataResponse> getMonthData({String? month}) async {
    final r = await _client.get(
      'monthData',
      params: {if (month != null) 'month': month},
    );
    return MonthDataResponse.fromJson(r);
  }

  Future<HistoricalSummaryResponse> getHistoricalSummary() async {
    final r = await _client.get('historicalSummary');
    return HistoricalSummaryResponse.fromJson(r);
  }

  Future<LastEntriesResponse> getLastEntries({int n = 10}) async {
    final r = await _client.get('lastEntries', params: {'n': n});
    return LastEntriesResponse.fromJson(r);
  }

  Future<MutationResponse> updateEntry(int row, UpdateEntryFields fields) async {
    final r = await _client.post('updateEntry', {
      'row': row,
      'fields': fields.toJson(),
    });
    return MutationResponse.fromJson(r);
  }

  Future<MutationResponse> deleteEntry(int row) async {
    final r = await _client.post('deleteEntry', {'row': row});
    return MutationResponse.fromJson(r);
  }
}
